### PR TITLE
[ fix #6643 ] Throw error if adding rewrite rule that is mutually defined with declared but not-yet-defined function

### LINKED ISSUE #6643: Rewrite rules are allowed in implicit mutual blocks
```agda
{-# OPTIONS --rewriting --confluence-check #-}
module Test45 where

open import Agda.Builtin.Equality
open import Agda.Builtin.Bool

{-# BUILTIN REWRITE _≡_ #-}

p : true ≡ false
q : true ≡ false
p = q
{-# REWRITE p #-}

q = refl
```

### PR DESCRIPTION
**Pull request by jespercockx**
Fixes #6643

### CODE REVIEW

File: src/full/Agda/TypeChecking/Rewriting.hs [hunk 1]
```diff
@@ -194,6 +194,15 @@ checkRewriteRule q = do
   -- for a type signature whose body has not been type-checked yet.
   when (isEmptyFunction $ theDef def) $
     typeError $ IllegalRewriteRule q BeforeFunctionDefinition
+  -- Issue 6643: Also check that there are no mututal definitions
+  -- that are not yet defined.
+  whenJustM (asksTC envMutualBlock) $ \ mb -> do
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Rewriting.hs]: NB: the `$` before `\` is no longer needed thanks to `BlockArguments`.

**Review by andreasabel — COMMENTED** [review]
Could this be a recoverable error (warning) instead, making the REWRITE pragma dead code?

**Comment by jespercockx** [comment]
> Could this be a recoverable error (warning) instead, making the REWRITE pragma dead code?

There are quite a few errors in checking a rewrite rule where we could "recover" by not adding the rewrite rule. I could prepare a separate PR (on top of this one) to turn all of them into warnings, but doing it here feels a bit ad-hoc since it requires some refactoring of the relevant code (`checkRewriteRule` will no longer always return a `RewriteRule`).

File: src/full/Agda/TypeChecking/Rewriting.hs (hunk 1 above)
Comment (jespercockx) [reply src/full/Agda/TypeChecking/Rewriting.hs]: Oh, that's some muscle memory that is going to take a while to unlearn :)

**Comment by jespercockx** [comment]
>  I could prepare a separate PR (on top of this one) to turn all of them into warnings

See https://github.com/agda/agda/pull/7343.

**Pull request closed without merging**